### PR TITLE
feat(key-auth): supporting key-auth plugin to get key from query string

### PR DESCRIPTION
**Pull request by wisdom-yzh**
### What this PR does / why we need it:

fixes issue #4421 

Adding a new `query` property to key-auth plugin, which can get key value from http query string

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/key-auth.lua [hunk 1]
```diff
@@ -32,6 +32,10 @@ local schema = {
             type = "string",
             default = "apikey",
         },
+        query = {
```
Comment (tokers) [inline apisix/plugins/key-auth.lua:35]: Should add some constraints to indicate that `query` and `header` are exclusive.

File: apisix/plugins/key-auth.lua [hunk 2]
```diff
@@ -84,6 +88,12 @@ end
 
 function _M.rewrite(conf, ctx)
     local key = core.request.header(ctx, conf.header)
+
+    if not key then
```
Comment (tokers) [inline apisix/plugins/key-auth.lua:92]: We should not introduce using order implicitly.

**Comment by tokers** [comment]
@wisdom-yzh From the changes I guess you tried to introduce an using order between the `header` and `query`, it will be somewhat obscure, we may treat them equally but set them as exclusive in jsonschema (e.g. by introducing the `oneOf` primitive).

**Comment by wisdom-yzh** [comment]
@tokers Accord to description of #4421, I guess @forgaoqiang wants to get key from both query string and headers, not exclusive. Maybe I can define one field instead of two fields, and check key from both query and headers?

**Comment by spacewander** [comment]
I think "header and query string" is OK, as jwt-auth plugin does it: https://github.com/apache/apisix/blob/07bfe5b9c03ad361cb71541eb02fd661f1323213/apisix/plugins/jwt-auth.lua#L154

File: t/plugin/key-auth.t [hunk 3]
```diff
@@ -300,7 +300,45 @@ passed
 
 
 
-=== TEST 11: valid consumer
+=== TEST 11: customize querystring
```
Comment (spacewander) [inline t/plugin/key-auth.t:303]: Please don't add the test in the middle of the other test. And we need to verify it via `GET /hello?xxx` in its own version of "valid consumer"

**Comment by wisdom-yzh** [comment]
@spacewander hi~I have changed strategy to "header and query string", and then rename field from "header" to "key_name“.

**Comment by spacewander** [comment]
@wisdom-yzh 
Err... We can't change the existing fields which will break the user's configuration. Your original solution is good enough, please bring it back.

**Comment by wisdom-yzh** [comment]
> @wisdom-yzh
> Err... We can't change the existing fields which will break the user's configuration. Your original solution is good enough, please bring it back.

ok, i'll restore the origin commit and fix the test case issue soon

File: t/plugin/key-auth.t [hunk 4]
```diff
@@ -309,3 +309,51 @@ Authorization: auth-one
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST13: valid consumer
```
Comment (spacewander) [inline t/plugin/key-auth.t:353]: need a space after TEST

**Review by spacewander — APPROVED** [review]

File: t/plugin/key-auth.t (hunk 4 above)
Comment (wisdom-yzh) [reply t/plugin/key-auth.t]: ok resolved

**Review by tokers — APPROVED** [review]

**Comment by forgaoqiang** [comment]
Great, this feature is what I needed, thanks

**Pull request merged**